cluster/service/cluster_service.go

```diff
@@ -65,6 +73,110 @@ func (c clusterService) CreateOrSaveClusterFromConfig(ctx context.Context) error
+	errInvalidURLGenerationMsg = "unable to generate '%s' URL from '%s' (expected an 'api' subdomain)"
+	errInvalidTypeMsg          = "invalid type of cluster: '%s' (expected 'OSO', 'OCP' or 'OSD')"
+)
+
+// validate checks if all data in the given cluster is valid, and fills the missing/optional URLs using the `APIURL`
+func validate(clustr *repository.Cluster) error {
+	if strings.TrimSpace(clustr.Name) == "" {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errEmptyFieldMsg, "name"))
+	}
+	err := validateURL(clustr.URL)
+	if err != nil {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errInvalidURLMsg, "API", clustr.URL, err))
+	}
+	// check other urls (console, logging, metrics)
+	for kind, urlStr := range map[string]*string{
+		"console": &clustr.ConsoleURL,
+		"logging": &clustr.LoggingURL,
+		"metrics": &clustr.MetricsURL} {
+		// check the url
+		if strings.TrimSpace(*urlStr) == "" {
+			switch kind {
+			case "console":
+				consoleURL, err := cluster.ConvertAPIURL(clustr.URL, "console", "console")
+				if err != nil {
+					return err
+				}
+				*urlStr = consoleURL
+			case "metrics":
+				metricsURL, err := cluster.ConvertAPIURL(clustr.URL, "metrics", "")
+				if err != nil {
+					return err
+				}
+				*urlStr = metricsURL
+			case "logging":
+				// This is not a typo; the logging host is the same as the console host in current k8s
+				loggingURL, err := cluster.ConvertAPIURL(clustr.URL, "console", "console")
+				if err != nil {
+					return err
+				}
+				*urlStr = loggingURL
+			}
+		} else if err := validateURL(*urlStr); err != nil {
+			// validate the URL
+			return errors.NewBadParameterErrorFromString(fmt.Sprintf(errInvalidURLMsg, kind, *urlStr, err))
+		}
+	}
+	// validate the cluster type
+	switch clustr.Type {
+	case cluster.OSD, cluster.OCP, cluster.OSO:
+		// ok
+	default:
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errInvalidTypeMsg, clustr.Type))
+	}
+	// validate other non empty fields
+	if strings.TrimSpace(clustr.AuthClientID) == "" {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errEmptyFieldMsg, "auth-client-id"))
+	}
+	if strings.TrimSpace(clustr.AuthClientSecret) == "" {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errEmptyFieldMsg, "auth-client-secret"))
+	}
+	if strings.TrimSpace(clustr.AuthDefaultScope) == "" {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errEmptyFieldMsg, "auth-client-default-scope"))
+	}
+	if strings.TrimSpace(clustr.SAToken) == "" {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errEmptyFieldMsg, "service-account-token"))
+	}
+	if strings.TrimSpace(clustr.SAUsername) == "" {
+		return errors.NewBadParameterErrorFromString(fmt.Sprintf(errEmptyFieldMsg, "service-account-username"))
+	}
+	if strings.TrimSpace(clustr.TokenProviderID) == "" {
+		// generated a value based on the ID of this cluster, so it's easier to track
+		clustr.ClusterID = uuid.NewV4()
```

Comment:
I don't think we can set ClusterID here. CreateOrSave() will try to update the cluster if it exists. And we can't change its ID. It will fail to update then. I guess we have to load the cluster by API URL here and if it exists then use this ID?

Reply:
good point. Let me run a test to verify what happens when an existing cluster is updated, but no `TokenProviderID` value is given

Reply:
addressed in 638ecb7